third-party/realdds/src/dds-device-broadcaster.cpp

```diff
@@ -4,45 +4,45 @@
 #include <realdds/dds-device-broadcaster.h>
 
 #include <realdds/dds-participant.h>
+#include <realdds/dds-publisher.h>
+#include <realdds/dds-topic-writer.h>
 #include <realdds/dds-utilities.h>
 #include <realdds/dds-guid.h>
-#include <realdds/topics/device-info/device-info-msg.h>
-#include <realdds/topics/device-info/deviceInfoPubSubTypes.h>
+#include <realdds/topics/flexible/flexible-msg.h>
+#include <realdds/topics/flexible/flexiblePubSubTypes.h>
 
 #include <librealsense2/utilities/easylogging/easyloggingpp.h>
-#include <fastdds/dds/domain/DomainParticipantFactory.hpp>
-#include <fastdds/dds/domain/DomainParticipant.hpp>
-#include <fastdds/dds/domain/DomainParticipantListener.hpp>
-#include <fastdds/dds/publisher/Publisher.hpp>
-#include <fastdds/dds/publisher/DataWriter.hpp>
-#include <fastdds/dds/topic/TypeSupport.hpp>
-#include <fastdds/dds/publisher/qos/PublisherQos.hpp>
-#include <fastdds/rtps/participant/ParticipantDiscoveryInfo.h>
+#include <librealsense2/utilities/json.h>
 
 #include <iostream>
 
 using namespace eprosima::fastdds::dds;
 using namespace realdds;
 
 // We want to know when readers join our topic
-class dds_device_broadcaster::dds_client_listener : public eprosima::fastdds::dds::DataWriterListener
+class dds_device_broadcaster::dds_client_listener : public realdds::dds_topic_writer //dds_topic_writer is a listener
 {
 public:
-    dds_client_listener( dds_device_broadcaster * owner )
-        : eprosima::fastdds::dds::DataWriterListener()
+    dds_client_listener( std::shared_ptr< dds_topic > const & topic,
+                         std::shared_ptr< dds_publisher > const & publisher,
+                         dds_device_broadcaster * owner )
+        : dds_topic_writer( topic, publisher )
         , _owner( owner )
     {
     }
 
+    std::atomic_bool _new_reader_joined = { false };  // Used to indicate that a new reader has joined for this writer
+
+protected:
     void on_publication_matched( eprosima::fastdds::dds::DataWriter * writer,
-                                 const eprosima::fastdds::dds::PublicationMatchedStatus & info ) override
+                                 eprosima::fastdds::dds::PublicationMatchedStatus const & info ) override
     {
         if( info.current_count_change == 1 )
         {
             // We send the work to the dispatcher to avoid waiting on the mutex here.
             _owner->_dds_device_dispatcher.invoke( [this]( dispatcher::cancellable_timer ) {
                 {
-                    std::lock_guard< std::mutex > lock( _owner->_new_client_mutex );
+                    std::lock_guard< std::mutex > lock( _owner->_new_client_mutex ); //TODO - mutxe locking needed here? Dispatcher needed here?
```

Comment:
Yes it is needed, a different thread is watching this elements..

Reply:
`_new_reader_joined` and `_owner->_trigger_msg_send` are atomic so I don't think we need to lock here, only to ensure that `_new_reader_joined` is the first we update, and a comment will suffice.

Also not sure that we need to `invoke` a dispatcher here, only in `_new_client_handler` lambda. here we just update 2 variables and notify the condition variable. Can be done in calling (DDS) thread.

Reply:
Could be that  the order of the variable set can let us drop the mutex and dispatcher, we will consider it

Reply:
Need the mutex, cannot remove it.
Checking the predicate (`_trigger_msg_sent` in this case) and moving the thread to the waiting queue is not atomic. The notify can happen in between and the condition variable doesn't have a memory for it. It will be lost and the thread remain waiting.